src/nbl/video/SOpenGLState.h

```diff
@@ -0,0 +1,173 @@
+#ifndef __NBL_S_OPENGL_STATE_H_INCLUDED__
+#define __NBL_S_OPENGL_STATE_H_INCLUDED__
+
+#include "nbl/core/IReferenceCounted.h"
+#include "COpenGLBuffer.h"
+#include "nbl/video/COpenGLRenderpassIndependentPipeline.h"
+#include "nbl/video/COpenGLPipelineLayout.h"
+#include "nbl/video/COpenGLDescriptorSet.h"
+#include "nbl/video/COpenGLComputePipeline.h"
+#include "nbl/video/IOpenGL_FunctionTable.h"
+#include "nbl/asset/ECommonEnums.h"
+#include <limits>
+
+namespace nbl {
+namespace video
+{
+
+struct SOpenGLState
+{
+    static inline constexpr uint32_t MAX_VIEWPORT_COUNT = 16u;
+
+    using buffer_binding_t = asset::SBufferBinding<const COpenGLBuffer>;
+
+    struct SVAO {
+        GLuint GLname;
+        uint64_t lastUsed;
+    };
+    struct HashVAOPair
+    {
+		COpenGLRenderpassIndependentPipeline::SVAOHash first = {};
+		SVAO second = { 0u,0ull };
+		//extra vao state being cached
+		std::array<buffer_binding_t, IGPUMeshBuffer::MAX_ATTR_BUF_BINDING_COUNT> vtxBindings;
+        buffer_binding_t idxBinding;
+        asset::E_INDEX_TYPE idxType; // not really state but i have to keep it somewhere
```

Comment:
are these things settable on the VAO even if the `separate_vertex_attribute` extension is around?

Reply:
vertex and index bindings? this is whole vao state basically and i think you know this XD
index type? no

Reply:
wait, `separate_vertex_attribute` im not sure i know what this is

Reply:
the extension that introduces "vulkan" like notions of separate attribute <format,stride,relative_offset,binding> and binding <buffer,offset>

Reply:
i thought its in core since 4.3 or something

Reply:
yes it is, and changing the binding on a VAO (the buffer which provides the data) is basically free

Reply:
so what is this question about? i dont get it

Reply:
the bound vertex attribute source buffers and index buffers should be part of the state, but separate from the VAO hash (if VAO has different buffers bound, then just change the buffers on the VAO without creating a new one in the LRU cache)

Reply:
@Crisspl done?